mautic/mautic
# Editor fixes

Please answer the following questions:

| Q             | A
| ------------- | ---
| Bug fix?      | Y
| New feature?  | N
| BC breaks?    | N
| Deprecations? | N
| Fixed issues  | https://github.com/mautic/mautic/issues/2026,  https://github.com/mautic/mautic/issues/2086, https://github.com/mautic/mautic/issues/2155, https://github.com/mautic/mautic/pull/2032, https://github.com/mautic/mautic/issues/1837, https://www.mautic.org/community/index.php//4923-landing-page-html-editor-doesn-t-work/p1

## Description
Froala adds some HTML which breaks the original HTML, so it must be returned to the original state before form save. I worked like this:

1. A before submit event triggered the clearing method.
2. The clearing method created a virtual iframe, pasted the Froala's HTML inside, a simple jQuery method selected the Froala styles.

The issue with this was that the pre-submit event was synchronous and if the HTML had many external JS and CSS in it, the iframe returned only the head part of the theme and forgot about the body. This resulted in a blank page/email.

This PR makes sure that the iframe is fully loaded before the clearing and fetching the cleared HTML back to the form textarea. The iframe load is asynchronous, so the pre-submit event technique couldn't be used.
A async callback method is defined as a form attr instead and if there is a form with this attr, the submit won't happen until the callback is called.

> What a nice thing would be if Froala cleared the styles which it adds itself without asking anyone!

Also, the editors were simplified in sense that the editor used for descriptions has only 3 buttons to format text. Then there `advanced-editor` in the DWT which has all buttons and the most powerful editor is `editor-basic-fullpage` which is used in the builders and it uses iframe for its content and allows to insert also CSS and JS in it.

This PR also changes the "change theme" warning which now shows up only if the user changes the content of selected theme when creating a new page/email. It will not show the warning if a theme is selected as the first thing when creating a new page/email.

## Steps to reproduce the bug (if applicable)
Try to insert this HTML to the landing page editor:
```
<!DOCTYPE html>
<html>
    <head>
        <meta charset="UTF-8" />
        <link href="https://fonts.googleapis.com/css?family=Lato:400,700" rel="stylesheet" type="text/css" />
        <title>Corso Come creare contenuti straordinari per il web</title>
        <meta name="viewport" content="initial-scale=1, width=device-width, maximum-scale=1, minimum-scale=1, user-scalable=no" />
        <script src="https://code.jquery.com/jquery-1.9.1.min.js" integrity="sha256-wS9gmOZBqsqWxgIVgA8Y9WcQOa7PgSIX+rPA0VL2rbQ=" crossorigin="anonymous"></script>
        <script src="https://code.jquery.com/ui/1.11.4/jquery-ui.min.js" integrity="sha256-xNjb53/rY+WmG+4L6tTl9m6PpqknWZvRt0rO1SRnJzw=" crossorigin="anonymous"></script>
        <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.2/js/bootstrap.min.js"></script>
        <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/html5shiv/3.6.2/html5shiv.min.js"></script>
        <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/classie/1.0.0/classie.min.js"></script>
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.css" />
        <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.2/css/bootstrap.min.css" />
    </head>
    <body>
        <div id="wrapper">
            HTML PAGE CODE
        </div>
    </body>
</html>
```
Save it and you should see a blank page without the `HTML PAGE CODE` text in it.

## Steps to test this PR
Apply this PR and either use the dev mode or build the prod assets. Refresh the page and try to save the same HTML again. The text should appear in the page preview now.

This PR modified the global method for submitting forms. All forms must still work.
